### ISSUE TITLE
android:versionCode appended with '8' if numeric, is '28' if specified as string

### ISSUE THREAD

**Issue by kokokenada** [meteor/meteor#7205]
In version Meteor 1.3.2.4, running OSX, building Android with "meteor build" CLI command.

In mobile-config.js, if you specify App.setPreference('android:versionCode', 3), the version reported in the  .apk file and the release-build/android/project/build/intermediates/manifests/full/release/AndroidManifest.xml, the android:versionCode is 38, not the expected 3.  If you specify the   App.setPreference('android:versionCode', '3'), (as string instead of a number) the output is 28.

**Comment by laosb** [comment]
That sounds strange.

**Comment by kokokenada** [comment]
Strange indeed.  When I ran into the problem, I searched and found forum post (https://forums.meteor.com/t/issue-wrong-android-versioncode-in-meteor-run-and-build/24015), so others (or at least another) have had this problem.

**Comment by joaopiopedreira** [comment]
I can confirm that this is happening to me too (Meteor 1.2.1 here), although it's not a blocker. Would be nice to have it sorted though.

João Pio Pedreira
/// Sent from a mobile device ///

**Comment by johnthepink** [comment]
This has been happening to me since at least November 2015. I am setting the build number in mobile-config.js. This is a screenshot from Hockey, the build number was 10, and it became 108 after a `meteor build`

![screenshot 2016-06-28 15 12 51](https://cloud.githubusercontent.com/assets/816517/16429039/ce185e26-3d42-11e6-8be2-bf2d298cd798.png)

**Comment by abernix** [comment]
I'm inclined to believe that this is a problem (or something happening in) Cordova itself as Meteor does [practically nothing](https://github.com/meteor/meteor/search?l=javascript&q=buildNumber+OR+android-versionCode&type=Code&utf8=%E2%9C%93) with the version code besides stick it into the `config.xml` that Cordova uses.

The typical way to set the `android-versionCode` (and `ios-CFBundleVersion`) is by setting `buildNumber` in the `mobile-config.js` (with `App.info`, not `setPreference`) before building the bundle.

I'm curious if any of you could attach the contents of your `.meteor/local/cordova-build/config.xml` file after a build to see what Cordova received from Meteor.

**Comment by jackkav** [comment]
This is fixed in the new cordova-lib update it appears, and can cause the android app install to fail to update from the old version.

**Comment by abernix** [comment]
@jackkav Thanks so much for following up on this and providing this information.

First, it appears that this was never really a "bug", but more of an intentional way to fix (something for) non-multiarch builds.  The change which is removing that functionality is here: https://github.com/apache/cordova-android/commit/dc579417fe32a5ea1cf466d308fc487902b6036b and you'll see some of the comments confirming that it's a breaking change.

Briefly and paraphrased from the other issue:
> It is best if you manage your `buildNumber` manually.  If you read the Cordova source code (https://github.com/apache/cordova-android/commit/dc579417fe32a5ea1cf466d308fc487902b6036b) you might understand better.  If you provide `buildNumber` yourself using Meteor 1.4.2 or higher (which was upgraded for Cordova 6.2) you will avoid all the extra changes that Cordova tries to intelligently make and it will just use your version.

I don't know if there is any good recommendation here aside from "you should set `buildNumber`".  As you'll see in that Cordova commit, there are still changes being made to the build number in some cases (i.e. multiplying by 10).  It seems like no matter what Meteor does, there may be problems if you're not setting it yourself. 

This issue (and the other) may be helpful enough for those experiencing problems, but I think the suggestion for the History.md might be a good idea.  Maybe the wording should be something along the lines of:

> Due to changes to how Cordova generates version numbers for iOS and Android apps, you may experience issues with apps updating on user devices.  To avoid this, consider managing your `buildNumber` yourself using `App.info('buildNumber', 'XXX');` in `mobile-config.js`.   There are additional considerations if you have been setting `android:versionCode` or `ios-CFBundleVersion` so see #7205 and #6978 for more information.

I don't know that this problem will affect everyone though and don't want to make this sound too scary.  Thoughts?

**Comment by jackkav** [comment]
Thanks @abernix for the follow-up. It will only affect those coming from <1.4.2 with no buildNumber manually set and android apps in production. I would imagine those few would appreciate the above warning, before they make the update to 1.4.2. As an aside it I would imagine the android app store will not accept publishing an apk update with a lower build number, but since we aren't using the app store for our production app that may explain why we are getting blocked by this issue while other aren't.

**Comment by hwillson** [comment]
Hi guys - I think adding that warning to the `History.md` makes sense (with a few slight modifications based on https://github.com/meteor/meteor/issues/7205#issuecomment-262755496). Should we move ahead with this to get this issue closed? Happy to submit the change if we're all in agreement. Thanks!

**Comment by abernix** [comment]
Thanks for reminding me (And offering), @hwillson.  I had forgot about this.

I opened #8390. :)
